plugins/Rockets/SnapshotTask.cpp

```diff
@@ -0,0 +1,188 @@
+    }
+
+    if (_params.volumeParams == nullptr)
+    {
+        _params.volumeParams = std::make_unique<VolumeParameters>(
+            engine.getParametersManager().getVolumeParameters());
+    }
+
+    if (_engine.supportsConcurrentSnapshots())
+    {
+        _camera = engine.createCamera();
+        _scene =
+            engine.createScene(*_params.animParams, *_params.geometryParams,
+                               *_params.volumeParams);
+
+        _renderer = engine.createRenderer(*_params.animParams,
+                                          *_params.renderingParams);
+
+        const auto &renderer = engine.getRenderer();
+        _renderer->setCurrentType(renderer.getCurrentType());
+        _renderer->clonePropertiesFrom(renderer);
+        if (_params.camera)
+        {
+            *_camera = *_params.camera;
+            _camera->setCurrentType(engine.getCamera().getCurrentType());
+            _camera->clonePropertiesFrom(engine.getCamera());
+        }
+        else
+            *_camera = engine.getCamera();
+
+        _scene->copyFrom(engine.getScene());
+    }
+    else
+    {
+        _camera = engine.getCameraPtr();
+        _renderer = engine.getRendererPtr();
+    }
+}
+
+ImageGenerator::ImageBase64 SnapshotFunctor::_renderSnapshot(
+    const size_t numAccumulations)
+{
+    const auto nameShort =
+        _params.name.empty() ? "" : " " + shortenString(_params.name);
+    const auto msg = "Render snapshot" + nameShort + "...";
+
+    const auto isStereo =
+        _camera->hasProperty("stereo") && _camera->getProperty<bool>("stereo");
+    const auto names = isStereo ? strings{"0L", "0R"} : strings{"default"};
+
+    std::vector<FrameBufferPtr> frameBuffers;
+    for (const auto &name : names)
+    {
+        frameBuffers.push_back(
+            _engine.createFrameBuffer(name, _params.size,
+                                      FrameBufferFormat::rgba_i8));
+    }
+
+    for (size_t i = 0; i < numAccumulations; i++)
+    {
+        for (auto &frameBuffer : frameBuffers)
+        {
+            _camera->setBufferTarget(frameBuffer->getName());
+            _camera->markModified(false);
+            _camera->commit();
+            _camera->resetModified();
+            _renderer->render(frameBuffer);
+            frameBuffer->incrementAccumFrames();
+        }
+
+        progress(msg, 1.f / numAccumulations,
+                 static_cast<float>(i) / numAccumulations);
+    }
+
+    return _imageGenerator.createImage(frameBuffers, _params.format,
+                                       _params.quality);
+}
+
+ImageGenerator::ImageBase64 SnapshotFunctor::_takeSnapshotConcurrent()
+{
+    _scene->commit();
+
+    _camera->updateProperty("aspect", double(_params.size.x) / _params.size.y);
+    _camera->commit();
+
+    if (_params.renderingParams)
+    {
+        _params.renderingParams->setSamplesPerPixel(1);
+        _params.renderingParams->setSubsampling(1);
```

Comment:
Doesn't it work using samples per pixels for OSPRay as well?

Reply:
It does but OSPRay uses accumulation frames instead to get a nice progress bar.